File: apk_monytix/app/src/main/java/com/example/apk/budget/BudgetViewModel.kt
package com.example.apk.budget

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.apk.auth.AuthService
import com.example.apk.config.Config
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

class BudgetViewModel(
    private val authService: AuthService = AuthService()
) : ViewModel() {
    
    private val client = OkHttpClient()
    private val gson = Gson()
    
    // Recommendations
    private val _recommendations = MutableStateFlow<List<BudgetRecommendation>>(emptyList())
    val recommendations: StateFlow<List<BudgetRecommendation>> = _recommendations.asStateFlow()
    
    private val _isRecommendationsLoading = MutableStateFlow(false)
    val isRecommendationsLoading: StateFlow<Boolean> = _isRecommendationsLoading.asStateFlow()
    
    private val _recommendationsError = MutableStateFlow<String?>(null)
    val recommendationsError: StateFlow<String?> = _recommendationsError.asStateFlow()
    
    // Committed Budget
    private val _committedBudget = MutableStateFlow<CommittedBudget?>(null)
    val committedBudget: StateFlow<CommittedBudget?> = _committedBudget.asStateFlow()
    
    private val _isCommittedLoading = MutableStateFlow(false)
    val isCommittedLoading: StateFlow<Boolean> = _isCommittedLoading.asStateFlow()
    
    // Commit Action
    private val _isCommitting = MutableStateFlow(false)
    val isCommitting: StateFlow<Boolean> = _isCommitting.asStateFlow()
    
    private val _commitError = MutableStateFlow<String?>(null)
    val commitError: StateFlow<String?> = _commitError.asStateFlow()
    
    // MARK: - API Methods
    
    fun loadRecommendations(month: String? = null) {
        viewModelScope.launch {
            _isRecommendationsLoading.value = true
            _recommendationsError.value = null
            
            try {
                // Wait for auth if needed or check if authenticated
                if (!authService.isAuthenticated()) {
                     // Could retry or verify if session loading
                     // For now, if current session is null, we try one more check
                     if (authService.getCurrentSession() == null) {
                         throw Exception("Not authenticated")
                     }
                }
                
                val session = authService.getCurrentSession()
                    ?: throw Exception("Not authenticated")
                
                withContext(Dispatchers.IO) {
                    val url = if (month != null) {
                        "${Config.apiBaseUrl}budget/recommendations?month=$month"
                    } else {
                        "${Config.apiBaseUrl}budget/recommendations"
                    }
                    
                    val request = Request.Builder()
                        .url(url)
                        .get()
                        .addHeader("Authorization", "Bearer ${session.accessToken}")
                        .build()
                    
                    val response = client.newCall(request).execute()
                    
                    if (!response.isSuccessful) {
                        throw IOException("Error ${response.code}")
                    }
                    
                    val body = response.body?.string() ?: throw IOException("Empty response")
                    val recommendationsResponse = gson.fromJson(body, BudgetRecommendationsResponse::class.java)
                    _recommendations.value = recommendationsResponse.recommendations.sortedByDescending { it.score }
                }
            } catch (e: Exception) {
                android.util.Log.e("BudgetViewModel", "Error loading recommendations", e)
                _recommendationsError.value = e.message ?: "Failed to load recommendations"
            } finally {
                _isRecommendationsLoading.value = false
            }
        }
    }
    
    fun loadCommittedBudget(month: String? = null) {
        viewModelScope.launch {
            _isCommittedLoading.value = true
            
            try {
                val session = authService.getCurrentSession()
                    ?: throw Exception("Not authenticated")
                
                withContext(Dispatchers.IO) {
                    val url = if (month != null) {
                        "${Config.apiBaseUrl}budget/committed?month=$month"
                    } else {
                        "${Config.apiBaseUrl}budget/committed"
                    }
                    
                    val request = Request.Builder()
                        .url(url)
                        .get()
                        .addHeader("Authorization", "Bearer ${session.accessToken}")
                        .build()
                    
                    val response = client.newCall(request).execute()
                    
                    if (response.isSuccessful) {
                        val body = response.body?.string() ?: "{}"
                        val committedResponse = gson.fromJson(body, CommittedBudgetResponse::class.java)
                        _committedBudget.value = committedResponse.budget
                    }
                }
            } catch (e: Exception) {
                android.util.Log.e("BudgetViewModel", "Error loading committed budget", e)
            } finally {
                _isCommittedLoading.value = false
            }
        }
    }
    
    fun commitToPlan(
        planCode: String,
        month: String? = null,
        goalAllocations: Map<String, Double>? = null,
        notes: String? = null
    ) {
        viewModelScope.launch {
            _isCommitting.value = true
            _commitError.value = null
            
            try {
                val session = authService.getCurrentSession()
                    ?: throw Exception("Not authenticated")
                
                withContext(Dispatchers.IO) {
                    val commitRequest = BudgetCommitRequest(
                        planCode = planCode,
                        month = month,
                        goalAllocations = goalAllocations,
                        notes = notes
                    )
                    
                    val json = gson.toJson(commitRequest)
                    val requestBody = json.toRequestBody("application/json".toMediaType())
                    
                    val request = Request.Builder()
                        .url("${Config.apiBaseUrl}budget/commit")
                        .post(requestBody)
                        .addHeader("Authorization", "Bearer ${session.accessToken}")
                        .addHeader("Content-Type", "application/json")
                        .build()
                    
                    val response = client.newCall(request).execute()
                    
                    if (!response.isSuccessful) {
                        val errorBody = response.body?.string() ?: "Unknown error"
                        throw IOException("Error ${response.code}: $errorBody")
                    }
                    
                    val body = response.body?.string() ?: throw IOException("Empty response")
                    val commitResponse = gson.fromJson(body, BudgetCommitResponse::class.java)
                    _committedBudget.value = commitResponse.budget
                    
                    // Reload recommendations
                    loadRecommendations(month)
                }
            } catch (e: Exception) {
                android.util.Log.e("BudgetViewModel", "Error committing budget", e)
                _commitError.value = e.message ?: "Failed to commit budget"
            } finally {
                _isCommitting.value = false
            }
        }
    }
}

File: ios_monytix/ios_monytix/budget/BudgetViewModel.swift
//
//  BudgetViewModel.swift
//  ios_monytix
//
//  Created by santosh on 05/01/26.
//

import Foundation
import SwiftUI
import Combine

@MainActor
class BudgetViewModel: ObservableObject {
    private let service: BudgetService
    
    // Recommendations
    @Published var recommendations: [BudgetRecommendation] = []
    @Published var isRecommendationsLoading = false
    @Published var recommendationsError: String?
    
    // Committed Budget
    @Published var committedBudget: CommittedBudget?
    @Published var isCommittedLoading = false
    @Published var committedError: String?
    
    // Variance
    @Published var variance: BudgetVariance?
    @Published var isVarianceLoading = false
    @Published var varianceError: String?
    
    // Commit Action
    @Published var isCommitting = false
    @Published var commitError: String?
    
    init(authService: AuthService) {
        self.service = BudgetService(authService: authService)
    }
    
    // MARK: - Recommendations
    
    func loadRecommendations(month: String? = nil) async {
        isRecommendationsLoading = true
        recommendationsError = nil
        defer { isRecommendationsLoading = false }
        
        do {
            recommendations = try await service.getRecommendations(month: month)
        } catch {
            recommendationsError = formatError(error)
            print("Error loading recommendations: \(error)")
        }
    }
    
    // MARK: - Committed Budget
    
    func loadCommittedBudget(month: String? = nil) async {
        isCommittedLoading = true
        committedError = nil
        defer { isCommittedLoading = false }
        
        do {
            committedBudget = try await service.getCommittedBudget(month: month)
        } catch {
            committedError = formatError(error)
            print("Error loading committed budget: \(error)")
        }
    }
    
    // MARK: - Commit
    
    func commitToPlan(planCode: String, month: String? = nil, goalAllocations: [String: Double]? = nil, notes: String? = nil) async -> Bool {
        isCommitting = true
        commitError = nil
        defer { isCommitting = false }
        
        do {
            committedBudget = try await service.commitBudget(
                planCode: planCode,
                month: month,
                goalAllocations: goalAllocations,
                notes: notes
            )
            
            // Reload recommendations to update UI
            await loadRecommendations(month: month)
            
            return true
        } catch {
            commitError = formatError(error)
            print("Error committing budget: \(error)")
            return false
        }
    }
    
    // MARK: - Variance
    
    func loadVariance(month: String? = nil) async {
        isVarianceLoading = true
        varianceError = nil
        defer { isVarianceLoading = false }
        
        do {
            variance = try await service.getVariance(month: month)
        } catch {
            varianceError = formatError(error)
            print("Error loading variance: \(error)")
        }
    }
    
    // MARK: - Helper Methods
    
    private func formatError(_ error: Error) -> String {
        if let budgetError = error as? BudgetError {
            return budgetError.errorDescription ?? "Unknown error"
        }
        return error.localizedDescription
    }
    
    func formatCurrency(_ amount: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencySymbol = "₹"
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: amount)) ?? "₹0"
    }
}

